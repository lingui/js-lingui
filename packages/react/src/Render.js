// @flow
import * as React from "react"
import PropTypes from "prop-types"

export type RenderProps = {
  render?: any,
  className?: string
}

type RenderComponentProps = {
  value: string | ?Array<any>
} & RenderProps

export default class Render extends React.Component<RenderComponentProps> {
  props: RenderComponentProps

  static contextTypes = {
    linguiDefaultRender: PropTypes.any
  }

  render() {
    const { className, value } = this.props
    let render = this.props.render || this.context.linguiDefaultRender

    if (render === null || render === undefined) {
      return value
    } else if (typeof render === "string") {
      // Built-in element: h1, p
      return React.createElement(render, { className }, value)
    }

    return React.isValidElement(render)
      ? // Custom element: <p className="lear' />
        React.cloneElement(render, {}, value)
      : // Custom component: ({ translation }) => <a title={translation}>x</a>
        React.createElement(render, { translation: value })
  }
}
