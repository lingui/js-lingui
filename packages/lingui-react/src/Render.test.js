// @flow
import React from 'react'
import { shallow } from 'enzyme'
import Render from './Render'

describe('Render', function () {
  it('should render just a text without wrapping element', function () {
    const span = shallow(<Render value="Just a text" />)
    expect(span).toMatchSnapshot()

    const withClass = shallow(<Render className="info" value="Just a text" />)
    expect(withClass).toMatchSnapshot()
  })

  it('should render custom element', function () {
    const builtin = shallow(<Render render="h1" value="Headline" />)
    expect(builtin).toMatchSnapshot()

    const element = shallow(<Render render={<h1 />} value="Headline" />)
    expect(element).toMatchSnapshot()
  })

  it('should render custom component', function () {
    const component = shallow(<Render value="Title" render={
      ({ translation }) => <a title={translation}>X</a>
    } />)
    expect(component.dive()).toMatchSnapshot()
  })

  it('should take default render element from context', function () {
    const h1 = <h1 />
    const context = {
      linguiDefaultRender: h1
    }

    const span = shallow(<Render value="Just a text" />, { context })
    expect(span).toMatchSnapshot()
  })
})
