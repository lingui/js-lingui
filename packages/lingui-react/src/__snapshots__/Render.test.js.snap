// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render should render custom component 1`] = `
<a
  title="Title"
>
  X
</a>
`;

exports[`Render should render custom element 1`] = `
<h1>
  Headline
</h1>
`;

exports[`Render should render custom element 2`] = `
<h1>
  Headline
</h1>
`;

exports[`Render should render just a text without wrapping element 1`] = `"Just a text"`;

exports[`Render should render just a text without wrapping element 2`] = `"Just a text"`;

exports[`Render should take default render element from context 1`] = `
<h1>
  Just a text
</h1>
`;
